README.md
# Weave AI

Weave AI is a collection of Flux controllers and CLI that manage the 
lifecycle of Large Language Models (LLMs) on Kubernetes.

**Weave AI CLI** aims to be the easiest way to onboard LLMs on Kubernetes, 
and the **Weave AI controllers** manage the lifecycle of the LLMs, including
training, serving, and monitoring of the models on production Kubernetes
clusters.

## Getting Started with Weave AI

Here's a step-by-step guide to run your first LLM with Weave AI.

### Prerequisites

Please install Kubernetes v1.27+ and Flux v2.1.0+ before proceeding.
Minimum requirements of the Kubernetes cluster are 8 CPUs and 16GB of memory with 100GB of SSD storage.

### Step 1: Install Weave AI CLI

First, install the Weave AI Command Line Interface (CLI). Use one of the following commands in your terminal:

#### Install with CURL

```shell
curl -s https://raw.githubusercontent.com/weave-ai/weave-ai/main/install/weave-ai.sh | sudo bash
````

You should see information about the download and installation process, confirming successful completion.

```shell
[INFO]  Downloading metadata https://api.github.com/repos/weave-ai/weave-ai/releases/latest
[INFO]  Using 0.11.0 as release
[INFO]  Downloading hash https://github.com/weave-ai/weave-ai/releases/download/v0.11.0/weave-ai_0.11.0_checksums.txt
[INFO]  Downloading binary https://github.com/weave-ai/weave-ai/releases/download/v0.11.0/weave-ai_0.11.0_linux_amd64.tar.gz
[INFO]  Verifying binary download
[INFO]  Installing weave-ai to /usr/local/bin/weave-ai
```

#### Or, Install with Homebrew
```shell
brew install weave-ai/tap/weave-ai
```

### Step 2: Create a KIND Cluster

Next, set up a Kubernetes in Docker (KIND) cluster with this command:

```shell
kind create cluster
```

The output will confirm the creation of your cluster, detailing the steps like node image preparation, nodes configuration, and control-plane initiation.

```shell
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.27.3) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
```

### Step 3: Install Flux

Now, install Flux to manage your cluster resources. Note: You need to disable the default network policy:

```shell
flux install --network-policy=false
```

The output will show the successful installation of various components in the flux-system namespace.

```shell
✚ generating manifests
✔ manifests build completed
► installing components in flux-system namespace
CustomResourceDefinition/alerts.notification.toolkit.fluxcd.io created
CustomResourceDefinition/buckets.source.toolkit.fluxcd.io created
CustomResourceDefinition/gitrepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/helmcharts.source.toolkit.fluxcd.io created
CustomResourceDefinition/helmreleases.helm.toolkit.fluxcd.io created
CustomResourceDefinition/helmrepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/kustomizations.kustomize.toolkit.fluxcd.io created
CustomResourceDefinition/ocirepositories.source.toolkit.fluxcd.io created
CustomResourceDefinition/providers.notification.toolkit.fluxcd.io created
CustomResourceDefinition/receivers.notification.toolkit.fluxcd.io created
Namespace/flux-system created
ResourceQuota/flux-system/critical-pods-flux-system created
ServiceAccount/flux-system/helm-controller created
ServiceAccount/flux-system/kustomize-controller created
ServiceAccount/flux-system/notification-controller created
ServiceAccount/flux-system/source-controller created
ClusterRole/crd-controller-flux-system created
ClusterRole/flux-edit-flux-system created
ClusterRole/flux-view-flux-system created
ClusterRoleBinding/cluster-reconciler-flux-system created
ClusterRoleBinding/crd-controller-flux-system created
Service/flux-system/notification-controller created
Service/flux-system/source-controller created
Service/flux-system/webhook-receiver created
Deployment/flux-system/helm-controller created
Deployment/flux-system/kustomize-controller created
Deployment/flux-system/notification-controller created
Deployment/flux-system/source-controller created
◎ verifying installation
✔ helm-controller: deployment ready
✔ kustomize-controller: deployment ready
✔ notification-controller: deployment ready
✔ source-controller: deployment ready
✔ install finished
```

### Step 4: Install Weave AI in Your Cluster

At this stage, you're ready to install Weave AI and its controller(s):

```shell
weave-ai install
```

After installation, you'll see confirmation messages, indicating that the Weave AI controllers and the default model catalog are set up.

```shell
✚ generating manifests
✔ manifests build completed
► installing components in weave-ai namespace
CustomResourceDefinition/languagemodels.ai.contrib.fluxcd.io created
Namespace/weave-ai created
Role/default/lm-tenant-role created
Role/weave-ai/lm-leader-election-role created
ClusterRole/lm-manager-role created
RoleBinding/default/lm-tenant-role-binding created
RoleBinding/weave-ai/lm-leader-election-rolebinding created
ClusterRoleBinding/lm-cluster-reconciler created
ClusterRoleBinding/lm-manager-rolebinding created
Deployment/weave-ai/lm-controller created
◎ verifying installation
✔ lm-controller: deployment ready
✔ install finished
```

### Step 5: Listing Models

To view the available models in your cluster, use:

```shell
weave-ai models
```

This command lists all OCI models, which are initially in an `INACTIVE` state to conserve resources.


```shell
NAME                               VERSION           FAMILY  STATUS    CREATED
weave-ai/dragon-yi-6b              v0.0.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/llama-2-7b-chat           v1.0.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/llama-2-7b-instruct-32k   v1.0.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/llamaguard-7b             v0.1.0-q4km-gguf          INACTIVE  1 minute ago
weave-ai/mistral-7b-instruct-v0.1  v0.1.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/mistral-7b-v0.1           v0.1.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/mistrallite-7b            v1.0.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/orca-2-7b                 v1.0.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/stablelm-zephyr-3b        v0.1.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/tinyllama-1.1b-chat       v0.3.0-q3ks-gguf          INACTIVE  1 minute ago
weave-ai/yarn-mistral-7b-128k      v0.1.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/zephyr-7b-alpha           v1.0.0-q5km-gguf          INACTIVE  1 minute ago
weave-ai/zephyr-7b-beta            v1.0.0-q5km-gguf          INACTIVE  1 minute ago
```

### Step 6: Create and run a model instance.

To activate and run a model, use the following command:

```shell
weave-ai run -d --ui --name my-model weave-ai/zephyr-7b-beta
```

This command activates the model and sets up a UI for interaction. Follow the instructions for port forwarding to access the LLM and the UI.

```shell
► checking if model weave-ai/zephyr-7b-beta exists and is active
► activate model weave-ai/zephyr-7b-beta
◎ waiting for model weave-ai/zephyr-7b-beta to be active
► creating new LLM instance default/my-model
◎ waiting for default/my-model to be ready
◎ waiting for default/my-model-chat-app to be ready
✔ to connect to your LLM:
  kubectl port-forward -n default svc/my-model 8000:8000
✔ to connect to the UI:
  kubectl port-forward -n default deploy/my-model-chat-app 8501:8501
```

Simply run the UI port-forward command.

```shell
kubectl port-forward -n default deploy/my-model-chat-app 8501:8501
```

Then open your browser to `http://localhost:8501` to try the model via our quick chat app.

![Chat UI](https://github.com/weave-ai/weave-ai/assets/10666/ff6e624e-90d5-42d9-9197-245619b1c4fa)

### Delete the LM instance.

Finally, to remove the LM instance and the associated UI, execute:

```shell
kubectl delete lm/my-model
```

This command deletes the specified language model instance from your cluster, also with the default chat UI if you've created one.
```
languagemodel.ai.contrib.fluxcd.io "my-model" deleted
```


CHANGELOG.md
# Changelog

All notable changes to this project are documented in this file.

## v0.12.0

**Release date:** 2023-12-28

- Update LM-controller to `v0.10.0`
- Update llama-cpp-python to `v0.2.23`
- Add labels and the service object to the UI
- Improve documentations for the `weave-ai run` command

## v0.11.0

**Release date:** 2023-12-13

- Add `stablelm-zephyr-3b` to the model catalog
- Fix `isActive` logic in the model activation process
- Reconcile the model in its activation process
- Add `--ui` flag to the `run` command to start Chat UI for the current model
- Add `llamaguard-7b` to the model catalog
- Add an instruction to install unreleased versions
- Update LM-controller to `v0.9.0`
- Update llama-cpp-python to `v0.2.22`
- Update chat-ui to `v0.2.0`

## v0.10.0

**Release date:** 2023-12-05

- Add `weave-ai activate-model` command

## Pre v0.10.0

- Add `weave-ai run` command
- Add `weave-ai models` command
- Initial release


cmd/weave-ai/run.go
package main

import (
	"context"
	"fmt"
	fluxmeta "github.com/fluxcd/pkg/apis/meta"
	"io"
	appsv1 "k8s.io/api/apps/v1"
	"k8s.io/apimachinery/pkg/util/intstr"
	"k8s.io/client-go/kubernetes"

	"github.com/weave-ai/weave-ai/pkg/utils"
	corev1 "k8s.io/api/core/v1"
	apimeta "k8s.io/apimachinery/pkg/api/meta"
	"k8s.io/apimachinery/pkg/api/resource"
	"k8s.io/apimachinery/pkg/util/wait"
	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
	"strings"
	"time"

	"github.com/spf13/cobra"
	aiv1a1 "github.com/weave-ai/lm-controller/api/v1alpha1"
	namesgenerator "github.com/weave-ai/weave-ai/pkg/namegenerator"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

var runCmd = &cobra.Command{
	Use:   "run",
	Args:  cobra.ExactArgs(1),
	Short: "Deploy and run an LLM",
	Long: `
# Deploy and run an LLM, e.g. zephyr-7b-beta, in the default namespace.
# The default model namespace is weave-ai.
weave-ai run zephyr-7b-beta

# Deploy and run an LLM, e.g. zephyr-7b-beta, in the default namespace as my-llm.
# The default model namespace is weave-ai.
weave-ai run --name=my-llm zephyr-7b-beta

# Deploy and run an LLM, e.g. zephyr-7b-beta from the weave-ai model namespace, in the default namespace.
weave-ai run weave-ai/zephyr-7b-beta

# Deploy and run an LLM, e.g. zephyr-7b-beta, in the default namespace and publish it as a LoadBalancer service.
weave-ai run -p -d weave-ai/zephyr-7b-beta

# Run zephyr-7b-beta using 6 CPU units.
weave-ai run -c 6 zephyr-7b-beta

# Run zephyr-7b-beta with 8 CPUs, detached, and named 'llm-test'.
weave-ai run -c 8 -d --name=llm-test zephyr-7b-beta

# Run zephyr-7b-beta in a custom namespace 'dev-space'.
weave-ai run -n dev-space zephyr-7b-beta

# Deploys zephyr-7b-beta in 'prod-space', as a LoadBalancer service with UI.
weave-ai run -n prod-space -p --ui zephyr-7b-beta

# Runs zephyr-7b-beta with 5 CPUs, in 'test-space', 
# detached, named 'llm-prod', published as a service with UI.
weave-ai run -c 5 -n test-space -d --name=llm-prod -p --ui zephyr-7b-beta
`,
	RunE: runCmdRun,
}

var runFlags struct {
	namespace      string
	name           string // name of the LLM
	publish        bool   // publish the LLM, which means it will be exposed as a LoadBalancer service
	cpu            string
	modelName      string
	modelNamespace string
	detach         bool // detach from the process e.g. not follow the logs
	ui             bool // start the UI
	local          bool // run the LLM locally
}

func init() {
	runCmd.Flags().StringVarP(&runFlags.cpu, "cpu", "c", "4", "specifies CPU resources for the LLM")
	runCmd.Flags().BoolVarP(&runFlags.detach, "detach", "d", false, "detaches from the Pod session, allowing the LLM to run in the background without showing logs")
	runCmd.Flags().StringVar(&runFlags.name, "name", "", "assigns a name to the LLM instance for identification")
	runCmd.Flags().BoolVarP(&runFlags.publish, "publish", "p", false, "makes the LLM available as a network-accessible LoadBalancer service")
	runCmd.Flags().BoolVar(&runFlags.ui, "ui", false, "starts the Weave Chat UI with the LLM for graphical interaction")
	// runCmd.Flags().BoolVar(&runFlags.local, "local", false, "run the LLM locally")

	// TODO use the default namespace from context
	runCmd.Flags().StringVarP(&runFlags.namespace, "namespace", "n", "default", "runs the LLM in the specific namespace")
	rootCmd.AddCommand(runCmd)
}

func runCmdRun(cmd *cobra.Command, args []string) error {
	if runFlags.local {
		return runCmdRunLocal(cmd, args)
	}
	return runCmdRun0(cmd, args)
}

func runCmdRunLocal(cmd *cobra.Command, args []string) error {
	/*
		modelName := args[0]
		// if model name contains / split it into model namespace and model name
		if strings.Contains(modelName, "/") {
			split := strings.SplitN(modelName, "/", 2)
			runFlags.modelNamespace = split[0]
			runFlags.modelName = split[1]
		} else {
			runFlags.modelName = modelName
			runFlags.modelNamespace = defaultNamespace
		}

		// 1. check ~/.weave-ai/bin
		// 2. download https://github.com/Mozilla-Ocho/llamafile/releases/download/0.4/llamafile-0.4
		// 3. rename to llamafile
		// 4. chmod +x llamafile
		// 5. move to ~/.weave-ai/bin
		// 6. run llamafile run
	*/
	return nil
}

func runCmdRun0(cmd *cobra.Command, args []string) error {
	modelName := args[0]
	// if model name contains / split it into model namespace and model name
	if strings.Contains(modelName, "/") {
		split := strings.SplitN(modelName, "/", 2)
		runFlags.modelNamespace = split[0]
		runFlags.modelName = split[1]
	} else {
		runFlags.modelName = modelName
		runFlags.modelNamespace = defaultNamespace
	}

	lmName := runFlags.name
	if lmName == "" {
		// random name using the docker name lib
		lmName = namesgenerator.GetRandomName(0)
	}

	serviceType := "ClusterIP"
	if runFlags.publish {
		serviceType = "LoadBalancer"
	}

	lm := &aiv1a1.LanguageModel{
		TypeMeta: metav1.TypeMeta{
			Kind:       "LanguageModel",
			APIVersion: "ai.contrib.fluxcd.io/v1alpha1",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name:      lmName,
			Namespace: runFlags.namespace,
			Labels: map[string]string{
				"ai.contrib.fluxcd.io/model-namespace": runFlags.modelNamespace,
				"ai.contrib.fluxcd.io/model":           runFlags.modelName,
			},
		},
		Spec: aiv1a1.LanguageModelSpec{
			SourceRef: aiv1a1.CrossNamespaceSourceReference{
				Kind:      "OCIRepository",
				Name:      runFlags.modelName,
				Namespace: runFlags.modelNamespace,
			},
			Interval:      metav1.Duration{Duration: 2 * time.Minute},
			RetryInterval: metav1.Duration{Duration: 30 * time.Second},
			Timeout:       &metav1.Duration{Duration: 2 * time.Minute},
			Prune:         true,
			Engine: aiv1a1.EngineSpec{
				ServiceType: corev1.ServiceType(serviceType),
				Replicas:    &[]int32{1}[0],
				Resources: corev1.ResourceRequirements{
					Requests: corev1.ResourceList{
						corev1.ResourceCPU: resource.MustParse(runFlags.cpu),
					},
				},
			},
		},
	}

	if createLmFlags.export {
		// TODO export manifests instead of installing
		// fmt.Print(buffer.String())
		return nil
	}

	ctx, cancelFn := context.WithTimeout(context.Background(), rootArgs.timeout)
	defer cancelFn()

	client, err := utils.KubeClient(kubeconfigArgs, kubeclientOptions)
	if err != nil {
		return err
	}

	if err := activateModel(ctx, client, runFlags.modelNamespace, runFlags.modelName, true); err != nil {
		return err
	}

	logger.Actionf("creating new LLM instance %s/%s", runFlags.namespace, lmName)
	if err := client.Create(ctx, lm); err != nil {
		return err
	}

	logger.Waitingf("waiting for %s/%s to be ready", runFlags.namespace, lmName)
	waitCtx, waitCancel := context.WithCancel(ctx)
	wait.UntilWithContext(waitCtx, func(ctx context.Context) {
		if err := client.Get(ctx, runtimeclient.ObjectKeyFromObject(lm), lm); err != nil {
			return
		}
		cond := apimeta.FindStatusCondition(lm.Status.Conditions, fluxmeta.ReadyCondition)
		if cond == nil {
			return
		}
		if cond.Status != metav1.ConditionTrue {
			return
		}
		waitCancel()
	}, 2*time.Second)

	svc := &corev1.Service{}
	if err := client.Get(ctx, runtimeclient.ObjectKeyFromObject(lm), svc); err != nil {
		return err
	}

	if runFlags.publish {
		logger.Waitingf("waiting for language model %s/%s to be published", runFlags.namespace, lmName)
		waitCtx, waitCancel := context.WithCancel(ctx)
		wait.UntilWithContext(waitCtx, func(ctx context.Context) {
			if err := client.Get(ctx, runtimeclient.ObjectKeyFromObject(svc), svc); err != nil {
				return
			}
			if len(svc.Status.LoadBalancer.Ingress) == 0 {
				return
			}
			if svc.Status.LoadBalancer.Ingress[0].IP == "" {
				return
			}
			waitCancel()
		}, 2*time.Second)
		ip := svc.Status.LoadBalancer.Ingress[0].IP
		logger.Successf("your LLM is ready at http://%s:8000", ip)
	}

	var (
		ui    *appsv1.Deployment
		uiSvc *corev1.Service
	)

	if runFlags.ui {
		uiAppName := lmName + "-chat-app"
		clusterDomain := rootArgs.clusterDomain
		ownerRefs := []metav1.OwnerReference{
			{
				APIVersion:         "ai.contrib.fluxcd.io/v1alpha1",
				Kind:               "LanguageModel",
				Name:               lm.Name,
				UID:                lm.UID,
				BlockOwnerDeletion: &[]bool{true}[0],
				Controller:         &[]bool{true}[0],
			},
		}

		labels := map[string]string{"app": uiAppName}
		ui = &appsv1.Deployment{
			TypeMeta: metav1.TypeMeta{
				APIVersion: "apps/v1",
				Kind:       "Deployment",
			},
			ObjectMeta: metav1.ObjectMeta{
				Name:            uiAppName,
				Namespace:       runFlags.namespace,
				Labels:          labels,
				OwnerReferences: ownerRefs,
			},
			Spec: appsv1.DeploymentSpec{
				Replicas: &[]int32{1}[0],
				Selector: &metav1.LabelSelector{
					MatchLabels: labels,
				},
				Template: corev1.PodTemplateSpec{
					ObjectMeta: metav1.ObjectMeta{
						Labels: labels,
					},
					Spec: corev1.PodSpec{
						SecurityContext: &corev1.PodSecurityContext{
							RunAsUser:    &[]int64{65532}[0],
							RunAsNonRoot: &[]bool{true}[0],
						},
						Containers: []corev1.Container{
							{
								Name:  "chat-app",
								Image: ImageChatInfo,
								Env: []corev1.EnvVar{
									{
										Name:  "LLM_API_HOST",
										Value: svc.Name + "." + runFlags.namespace + ".svc." + clusterDomain + ":8000",
									},
								},
								SecurityContext: &corev1.SecurityContext{
									Privileged:   &[]bool{false}[0],
									RunAsNonRoot: &[]bool{true}[0],
									RunAsUser:    &[]int64{65532}[0],
									Capabilities: &corev1.Capabilities{
										Drop: []corev1.Capability{
											"ALL",
										},
									},
								},
								Ports: []corev1.ContainerPort{
									{
										ContainerPort: 8501,
										Name:          "http",
										Protocol:      corev1.ProtocolTCP,
									},
								},
							},
						},
					},
				},
			},
		}
		if err := client.Create(ctx, ui); err != nil {
			return err
		}

		uiSvc = &corev1.Service{
			TypeMeta: metav1.TypeMeta{
				APIVersion: "v1",
				Kind:       "Service",
			},
			ObjectMeta: metav1.ObjectMeta{
				Name:            uiAppName,
				Namespace:       runFlags.namespace,
				Labels:          labels,
				OwnerReferences: ownerRefs,
			},
			Spec: corev1.ServiceSpec{
				Selector: labels,
				Type:     corev1.ServiceTypeClusterIP,
				Ports: []corev1.ServicePort{
					{
						Name:       "http",
						Port:       8501,
						TargetPort: intstr.FromInt32(8501),
					},
				},
			},
		}

		if err := client.Create(ctx, uiSvc); err != nil {
			return err
		}

		// wait is good for the UI to be ready
		logger.Waitingf("waiting for %s/%s to be ready", runFlags.namespace, uiAppName)
		waitCtx, waitCancel := context.WithCancel(ctx)
		wait.UntilWithContext(waitCtx, func(ctx context.Context) {
			if err := client.Get(ctx, runtimeclient.ObjectKeyFromObject(ui), ui); err != nil {
				return
			}
			var cond *appsv1.DeploymentCondition
			for _, condition := range ui.Status.Conditions {
				if condition.Type == appsv1.DeploymentAvailable {
					cond = &condition
					break
				}
			}
			if cond == nil {
				return
			}
			if cond.Status != corev1.ConditionTrue {
				return
			}
			waitCancel()
		}, 2*time.Second)

	}

	if !runFlags.detach {
		pods := &corev1.PodList{}

		// engine pods
		matchingLabels := runtimeclient.MatchingLabels{"app": lm.Name}
		podLogOpts := corev1.PodLogOptions{
			Container: "engine",
			Follow:    true,
		}

		// if UI is enabled, wait for the UI pod to be ready
		if runFlags.ui {
			matchingLabels["app"] = lm.Name + "-chat-app"
			podLogOpts.Container = "chat-app"
		}

		if err := client.List(ctx, pods,
			runtimeclient.InNamespace(runFlags.namespace),
			matchingLabels,
		); err != nil {
			return err
		}

		podName := pods.Items[0].Name
		config, err := utils.KubeConfig(kubeconfigArgs, kubeclientOptions)
		if err != nil {
			return err
		}

		clientSet, err := kubernetes.NewForConfig(config)
		if err != nil {
			return err
		}
		req := clientSet.CoreV1().Pods(runFlags.namespace).GetLogs(podName, &podLogOpts)

		var podLogs io.ReadCloser
		i := 0
		const MaxRetry = 10
		for {
			var err error
			podLogs, err = req.Stream(context.Background())
			if err != nil {
				time.Sleep(1 * time.Second)
				i = i + 1
				if i > MaxRetry {
					return err
				}
			}
			break
		}
		if podLogs == nil {
			return fmt.Errorf("could not get logs for pod %s/%s", runFlags.namespace, podName)
		}
		defer podLogs.Close()

		// Stream the pod logs
		buf := make([]byte, 2000)
		for {
			w, err := podLogs.Read(buf)
			if w > 0 {
				fmt.Print(string(buf[:w]))
			}
			if err != nil {
				break
			}
		}
	} else {
		// if detached, shows kubectl port-forward commands
		logger.Successf("to connect to your LLM:\n  kubectl port-forward -n %s svc/%s 8000:8000", svc.Namespace, svc.Name)
		if runFlags.ui {
			logger.Successf("to connect to the UI:\n  kubectl port-forward -n %s svc/%s 8501:8501", uiSvc.Namespace, uiSvc.Name)
		}
	}

	return nil
}
